staging/src/k8s.io/apiserver/pkg/server/filters/with_early_late_annotations.go

```diff
@@ -64,16 +65,22 @@ func exemptIfHealthProbe(r *http.Request) bool {
 	return false
 }
 
-// WithShutdownLateAnnotation, if added to the handler chain, tracks the
-// incoming request(s) after the apiserver has initiated the graceful
-// shutdown, and annoates the audit event for these request(s) with
-// diagnostic information.
-// This enables us to identify the actor(s)/load balancer(s) that are sending
-// requests to the apiserver late during the server termination.
-// It should be placed after (in order of execution) the
-// 'WithAuthentication' filter.
-func WithShutdownLateAnnotation(handler http.Handler, shutdownInitiated lifecycleEvent, delayDuration time.Duration) http.Handler {
-	return withShutdownLateAnnotation(handler, shutdownInitiated, delayDuration, exemptIfHealthProbe, clockutils.RealClock{})
+// WithShutdownResponseHeader, if added to the handler chain, adds a header
+// 'X-OpenShift-Disruption' to the response with the following information:
+//
+//	shutdown={true|false} shutdown-delay-duration=%s elapsed=%s host=%s
+//	 shutdown: whether the server is shutting down gracefully.
+//	 shutdown-delay-duration: value of --shutdown-delay-duration server run option
+//	 elapsed: how much time elapsed since the server received the TERM signal
+//	 host: host name of the server, it is used to identify the server instance
+//	       from the others.
+//
+// This handler will add the response header only if the client opts in by
+// adding the 'X-Openshift-If-Disruption' header in the request.
+func WithShutdownResponseHeader(handler http.Handler, shutdownInitiated lifecycleEvent, delayDuration time.Duration) http.Handler {
+	return withShutdownResponseHeader(handler, shutdownInitiated, delayDuration, func() string {
+		return os.Getenv("HOSTNAME")
```

Comment:
question: is host name a sensitive information that should never leave the cluster?

Reply:
Is this environment variable guaranteed to be set? In-tree I see mostly `os.Hostname()` for this.

If you switch, you'd probably want to avoid the extra syscalls on every request -- elsewhere it looks like we assume the return value never changes during the life of the process.

If we're concerned about sharing hostname with clients, can we use the ID from https://github.com/kubernetes/enhancements/blob/master/keps/sig-api-machinery/1965-kube-apiserver-identity/? It's essentially a hash of the hostname, and tests should be able to correlate this by looking up the `kubernetes.io/hostname` label on the appropriate lease.

Reply:
> If you switch, you'd probably want to avoid the extra syscalls on every request -- elsewhere it looks like we assume the return value never changes during the life of the process.

This PR retrieved the host name once, when the handler chain was constructed, relying on the notion that the host name never changes during the life of the server process.

Yes, we are going to use APIServerIdentity, given that leaking the plain host name will be unacceptable to the customers, (I had hoped that we could avoid this additional complexity of the tests needing to correlate to the host name from the leases object, but now it's a necessary complexity)